vmm/src/sigsys_handler.rs

```diff
@@ -35,17 +41,66 @@ extern "C" fn sigsys_handler(
     if num != libc::SIGSYS {
         return;
     }
-    let syscall = unsafe { *(info as *const i32).offset(SI_OFF_SYSCALL) as usize };
-    METRICS.seccomp.num_faults.inc();
-    METRICS.seccomp.bad_syscalls[syscall].inc();
+
+    // We're also trying to log something here, which might not be re-entrant (now, or at some
+    // point in the future), so we make sure this logic only runs once. If another signal causes
+    // the handler to run again, we simply exit the process with the appropriate error code. This
+    // should be extremely unlikely to happen.
+    if HANDLING_BAD_SYSCALL.compare_and_swap(false, true, Ordering::SeqCst) == false {
+        let syscall = unsafe { *(info as *const i32).offset(SI_OFF_SYSCALL) as usize };
+        METRICS.seccomp.num_faults.inc();
+        METRICS.seccomp.bad_syscalls[syscall].inc();
+        error!("Shutting down VM after intercepting a bad syscall.");
```

Comment:
I don't think we should log here. Andreea is right that `bad_syscalls` should go away and the offender should be logged, but `fmt` ends up `extend`-ing a `Vec` (`fmt`-ing a number is safe because it uses a [static buffer](https://github.com/rust-lang/rust/blob/588f94b37d020aefe54b0c00ced2fdd7ccc61dfb/src/libcore/fmt/num.rs#L56) but `error!` formats the number into the string, then adds a whole bunch of other info from our `logger` impl - timestamp etc which eventually translates into a [realloc](https://github.com/rust-lang/rust/blob/588f94b37d020aefe54b0c00ced2fdd7ccc61dfb/src/liballoc/raw_vec.rs#L667)) so we shouldn't be using is inside a signal handler even if it's only for `SIGSYS`.

Can we store the syscall number in an `Atomic` instead and notify the vmm that it should log it first (thus deferring the pesky allocation to outside the signal handler), then end the process?